proof/refine/ARM/IpcCancel_R.thy

```diff
@@ -507,39 +507,1153 @@ locale delete_one = delete_one_conc + delete_one_abs +
+    apply (frule (2) heap_ls_next_in_list)
+    apply (simp add: sc_with_reply_def the_pred_option_def
+              split: if_split_asm)
+    apply blast
+   apply (rule conjI)
+    apply (clarsimp simp add: ghost_relation_def)
+    apply (erule_tac x=scp in allE)+
+    apply (clarsimp simp: obj_at_def a_type_def
+                   split: Structures_A.kernel_object.splits if_split_asm)
+   apply (clarsimp simp: caps_of_state_after_update cte_wp_at_after_update
+                         swp_def fun_upd_def obj_at_simps)
+   done
+qed
+
+end
+
+(* end : related corres rules *)
+
+(* FIXME RT: the original in Invariants_AI; do we want valid_objs or valid_tcbs? *)
+lemma st_tcb_at_valid_st2:
+  "\<lbrakk> st_tcb_at ((=) st) t s; valid_tcbs s \<rbrakk> \<Longrightarrow> valid_tcb_state st s"
+  apply (clarsimp simp add: valid_tcbs_def get_tcb_def pred_tcb_at_def obj_at_def)
+  apply (simp add: valid_obj_def valid_tcb_def)
+  done
+
+(* FIXME RT: original in DetSchedInvs_AI; maybe combine and clean up *)
+lemma BlockedOnReply_reply_tcb_reply_at:
+  " st_tcb_at ((=) (Structures_A.BlockedOnReply rptr)) tptr s
+   \<Longrightarrow> sym_refs (state_refs_of s)
+   \<Longrightarrow> reply_tcb_reply_at ((=) (Some tptr)) rptr s"
+  apply (clarsimp simp: pred_tcb_at_def)
+  apply (drule (1) sym_refs_obj_atD[where p=tptr])
+  apply (clarsimp simp: state_refs_of_def obj_at_def tcb_st_refs_of_def
+                 split: Structures_A.thread_state.splits)
+  apply (rename_tac ko; case_tac ko; clarsimp simp: get_refs_def2 reply_tcb_reply_at_def obj_at_def)
+  done
+
+(* FIXME RT: move to StateRelation.thy? Replace the non-projection one? *)
+lemma sc_replies_relation_scReply_cross:
+  "\<lbrakk> sc_replies_relation s s'; pspace_relation (kheap s) (ksPSpace s');
+     pspace_aligned s; pspace_distinct s; valid_objs s;
+   \<exists>n. kheap s ptr = Some (kernel_object.SchedContext sc n)\<rbrakk>
+    \<Longrightarrow> scReplies_of s' ptr = hd_opt (sc_replies sc)"
+  apply (prop_tac "sc_at' ptr s'")
+   apply (fastforce intro!: sc_at_cross valid_sched_context_size_objsI simp: obj_at_def is_sc_obj_def)
+  apply (clarsimp simp: sc_replies_relation_def obj_at'_def projectKOs)
+  apply (drule_tac x=ptr and y="sc_replies sc" in spec2)
+  apply (prop_tac "sc_replies_of s ptr = Some (sc_replies sc)")
+   apply (clarsimp simp: vs_heap_simps)
+  apply (clarsimp simp: sc_of_def opt_map_def)
+  apply (case_tac "sc_replies sc"; clarsimp simp: projectKO_opt_sc)
+  done
+
+(* wp rules *)
+
+lemma scReply_update_empty_sc_with_reply':
+  "\<lbrace>(\<lambda>s. sc_with_reply' rp s = Some scp) and ko_at' sc' scp
+   and (\<lambda>s. sym_refs (state_refs_of' s)) and reply_at' rp
+   and (\<lambda>s. sym_refs (list_refs_of_replies' s))
+   and K (scReply sc' = Some rp)
+   and pspace_aligned' and pspace_distinct'\<rbrace>
+    setSchedContext scp (scReply_update Map.empty sc')
+             \<lbrace>\<lambda>rv s. sc_with_reply' rp s = None\<rbrace>"
```

Comment:
Could you either put this on one line or align the function with the pre/postcondition? The function should be one space to the left, too